firedrake/functionspacedata.py

```diff
@@ -89,68 +91,72 @@ def get_node_set(mesh, nodes_per_entity):
     dm = PETSc.DMShell().create(mesh.comm)
     dm.setPointSF(mesh._plex.getPointSF())
     dm.setDefaultSection(global_numbering)
-    node_classes = tuple(numpy.dot(nodes_per_entity, mesh._entity_classes))
+    node_classes = mesh.node_classes(nodes_per_entity)
     node_set = op2.Set(node_classes, halo=halo_mod.Halo(dm), comm=mesh.comm)
     # Don't need it any more, explicitly destroy.
     dm.destroy()
-    extruded = bool(mesh.layers)
+    extruded = mesh.cell_set._extruded
     if extruded:
-        node_set = op2.ExtrudedSet(node_set, layers=mesh.layers)
+        # FIXME! This is a LIE! But these sets should not be extruded
+        # anyway, only the code gen in PyOP2 is busted.
+        node_set = op2.ExtrudedSet(node_set, layers=2)
```

Comment:
See [this](https://github.com/firedrakeproject/firedrake/compare/node-set-never-extruded) and [this](https://github.com/OP2/PyOP2/compare/fix-extruded-set-ugly).